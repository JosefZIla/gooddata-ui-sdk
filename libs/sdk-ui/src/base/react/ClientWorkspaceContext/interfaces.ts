// (C) 2019-2022 GoodData Corporation

/**
 * Resolved LCM identifiers of the workspace.
 *
 * @alpha
 */
export interface IClientWorkspaceIdentifiers {
    /**
     * Data product identifier.
     */
    dataProduct?: string;

    /**
     * Client identifier.
     */
    client?: string;

    /**
     * Segment identifier.
     */
    segment?: string;

    /**
     * Workspace identifier.
     */
    workspace?: string;
}

/**
 * @internal
 */
export interface IClientWorkspaceStatus {
    isInitialized: boolean;
}
