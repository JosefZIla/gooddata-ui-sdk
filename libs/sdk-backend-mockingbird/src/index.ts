// (C) 2019-2021 GoodData Corporation
/**
 * @packageDocumentation
 */
import { dummyBackend, dummyBackendEmptyData, dummyDataView } from "@gooddata/sdk-backend-base";
export { dummyBackend, dummyBackendEmptyData, dummyDataView };

export { recordedBackend } from "./recordedBackend";

export {
    RecordingIndex,
    ExecutionRecording,
    InsightRecording,
    DisplayFormRecording,
    ScenarioRecording,
    RecordedBackendConfig,
    CatalogRecording,
    VisClassesRecording,
    DashboardRecording,
    RecordedRefType,
    SecuritySettingsUrlValidator,
    SecuritySettingsOrganizationScope,
} from "./recordedBackend/types";

export {
    recordedDataView,
    recordedDataViews,
    recordedInsight,
    DataViewAll,
    dataViewWindow,
    DataViewFirstPage,
    NamedDataView,
} from "./recordedBackend/execution";

export { CompositeBackendPart, compositeBackend } from "./compositeBackend/index";

export {
    LegacyExecutionRecording,
    legacyRecordedBackend,
    legacyRecordedDataView,
    LegacyRecordingIndex,
    LegacyWorkspaceRecordings,
} from "./legacyRecordedBackend";
