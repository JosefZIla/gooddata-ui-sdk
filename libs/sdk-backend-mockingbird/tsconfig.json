{
    "compilerOptions": {
        "esModuleInterop": true,
        "target": "es5",
        "module": "commonjs",
        "moduleResolution": "node",
        "lib": ["es6", "dom", "ES2016.Array.Include"],
        "importHelpers": true,
        "declaration": true,
        "declarationMap": true,
        "sourceMap": true,
        "outDir": "./dist",
        "strict": true,
        "noImplicitAny": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitReturns": true,
        "baseUrl": "..",
        "paths": {
            "@gooddata/sdk-model": ["sdk-model/src"],
            "@gooddata/sdk-backend-base": ["sdk-backend-base/src"],
            "@gooddata/sdk-backend-spi": ["sdk-backend-spi/src"]
        }
    },
    "exclude": ["node_modules", "dist"]
}
