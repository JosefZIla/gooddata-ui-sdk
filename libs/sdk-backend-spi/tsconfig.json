{
    "compilerOptions": {
        "esModuleInterop": true,
        "target": "es2017",
        "module": "commonjs",
        "lib": ["es2017", "webworker"],
        "moduleResolution": "node",
        "importHelpers": true,
        "sourceMap": true,
        "outDir": "./dist",
        "strict": true,
        "noImplicitAny": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true,
        "noImplicitReturns": true,
        "removeComments": false,
        "baseUrl": "..",
        "paths": {
            "@gooddata/sdk-model": ["sdk-model/src"]
        }
    },
    "exclude": ["**/node_modules", "dist", "esm"]
}
