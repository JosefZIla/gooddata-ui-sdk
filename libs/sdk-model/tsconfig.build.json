{
    "extends": "./tsconfig.json",
    "include": ["src/index.ts", "src/dashboard/drillUrl.ts"],
    "compilerOptions": {
        "baseUrl": null,
        "paths": null,
        "declaration": true,
        "declarationMap": true
    }
}
