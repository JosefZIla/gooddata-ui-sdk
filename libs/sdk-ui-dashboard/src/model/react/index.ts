// (C) 2021-2022 GoodData Corporation
export {
    DashboardStoreProvider,
    useDashboardDispatch,
    useDashboardSelector,
    ReactDashboardContext,
} from "./DashboardStoreProvider";
export { IDashboardEventsContext, useDashboardEventsContext } from "./DashboardEventsContext";
export { useDashboardCommandProcessing, CommandProcessingStatus } from "./useDashboardCommandProcessing";
export { useDashboardEventDispatch } from "./useDashboardEventDispatch";
export {
    useDashboardQueryProcessing,
    QueryProcessingStatus,
    QueryProcessingErrorState,
    QueryProcessingPendingState,
    QueryProcessingRejectedState,
    QueryProcessingRunningState,
    QueryProcessingState,
    QueryProcessingSuccessState,
    UseDashboardQueryProcessingResult,
} from "./useDashboardQueryProcessing";
export { useDashboardUserInteraction } from "./useDashboardUserInteraction";
export { useDashboardAsyncRender, UseDashboardAsyncRender } from "./useDashboardAsyncRender";
export { IDashboardStoreProviderProps } from "./types";
export { useDispatchDashboardCommand } from "./useDispatchDashboardCommand";
export { useWidgetExecutionsHandler } from "./useWidgetExecutionsHandler";
export { useDashboardScheduledEmails } from "./useDashboardScheduledEmails";
