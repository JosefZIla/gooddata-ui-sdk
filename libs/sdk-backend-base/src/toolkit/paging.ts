// (C) 2019-2021 GoodData Corporation
import { IPagedResource } from "@gooddata/sdk-backend-spi";
import invariant from "ts-invariant";

/**
 * This implementation of {@link @gooddata/sdk-backend-spi#IPagedResource} pages over a list of items
 * provided at construction time. The paging is done using pre-configured page limit and starts at particular offset.
 *
 * @internal
 */
export class InMemoryPaging<T> implements IPagedResource<T> {
    public readonly items: T[];
    public readonly limit: number;
    public readonly offset: number;
    public readonly totalCount: number;

    constructor(protected readonly allItems: T[], limit = 50, offset = 0) {
        invariant(offset >= 0, `paging offset must be non-negative, got: ${offset}`);
        invariant(limit > 0, `limit must be a positive number, got: ${limit}`);

        // this will naturally return empty items if at the end of data; limit will always be positive
        this.items = allItems.slice(offset, offset + limit);

        // offset is at most at the end of all available elements
        this.offset = Math.min(offset, allItems.length);
        // limit is always kept as-requested
        this.limit = limit;

        this.totalCount = allItems.length;
    }

    public async next(): Promise<IPagedResource<T>> {
        if (this.items.length === 0) {
            return this;
        }

        return new InMemoryPaging(this.allItems, this.limit, this.offset + this.items.length);
    }

    public async goTo(pageIndex: number): Promise<IPagedResource<T>> {
        if (this.items.length === 0) {
            return this;
        }

        return new InMemoryPaging(this.allItems, this.limit, pageIndex * this.items.length);
    }

    public async all(): Promise<T[]> {
        return [...this.allItems];
    }

    public async allSorted(compareFn: (a: T, b: T) => number): Promise<T[]> {
        return [...this.allItems].sort(compareFn);
    }
}

/**
 * Given a paged result, this function will retrieve all pages from the backend concatenated to a single array.
 *
 * @param pagedResource - the paged resource to get all the pages of
 */
async function getAllPagesOfInner<T>(
    pagedResource: Omit<IPagedResource<T>, "all" | "allSorted">,
): Promise<T[]> {
    const results: T[] = [];
    const pageSize = pagedResource.limit;
    // if the paged resource is already at the 0 offset, use it directly to save a duplicate request
    let currentPage = pagedResource.offset !== 0 ? await pagedResource.goTo(0) : pagedResource;

    // eslint-disable-next-line no-constant-condition
    while (true) {
        results.push(...currentPage.items);

        if (currentPage.items.length < pageSize) {
            return results;
        }
        currentPage.items.sort();

        currentPage = await currentPage.next();
    }
}

/**
 * Given a paged result, this function will enhance it with the `all` implementation.
 *
 * @remarks TODO: FET-847 avoid the need for this function
 *
 * @param pagedResource - paged resource to enhance
 * @internal
 */
export function enhanceWithAll<TItem, TResource extends Omit<IPagedResource<TItem>, "all" | "allSorted">>(
    pagedResource: TResource,
): TResource & IPagedResource<TItem> {
    return {
        ...pagedResource,
        all: () => getAllPagesOfInner(pagedResource),
        allSorted: (compareFn) => getAllPagesOfInner(pagedResource).then((items) => items.sort(compareFn)),
    };
}
