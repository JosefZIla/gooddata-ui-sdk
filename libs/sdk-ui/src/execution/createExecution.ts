// (C) 2019-2021 GoodData Corporation
import {
    attributeLocalId,
    DimensionItem,
    IAttribute,
    IAttributeOrMeasure,
    IDimension,
    INullableFilter,
    isAttribute,
    ISortItem,
    ITotal,
    MeasureGroupIdentifier,
    newDimension,
    newTwoDimensional,
} from "@gooddata/sdk-model";
import { IAnalyticalBackend, IPreparedExecution } from "@gooddata/sdk-backend-spi";
import isEmpty from "lodash/isEmpty";
import invariant from "ts-invariant";

/**
 * @internal
 */
export type CreateExecutionOptions = {
    /**
     * Backend to execute against.
     *
     * Note: the backend must come either from this property or from BackendContext. If you do not specify
     * backend here, then the executor MUST be rendered within an existing BackendContext.
     */
    backend?: IAnalyticalBackend;

    /**
     * Workspace in whose context to perform the execution.
     *
     * Note: the workspace must come either from this property or from WorkspaceContext. If you do not specify
     * workspace here, then the executor MUST be rendered within an existing WorkspaceContext.
     */
    workspace?: string;

    /**
     * Data series will be built using the provided measures that are optionally further scoped for
     * elements of the specified attributes.
     */
    seriesBy: IAttributeOrMeasure[];

    /**
     * Optionally slice all data series by elements of these attributes.
     */
    slicesBy?: IAttribute[];

    /**
     * Optionally include these totals among the data slices.
     */
    totals?: ITotal[];

    /**
     * Optional filters to apply on server side.
     */
    filters?: INullableFilter[];

    /**
     * Optional sorting to apply on server side.
     */
    sortBy?: ISortItem[];

    /**
     * Optional informative name of the component. This value is sent as telemetry information together
     * with the actual execution request. We recommend to set this because it can be useful for diagnostic
     * purposes.
     *
     * Defaults 'Execute'.
     */
    componentName?: string;
};

/**
 * When caller desires just data series and no slicing, create a single-dim result.
 */
function seriesOnlyDim(seriesBy: IAttributeOrMeasure[]): IDimension[] {
    return [newDimension(seriesBy.filter(isAttribute).map(attributeLocalId).concat(MeasureGroupIdentifier))];
}

/**
 * When caller desires data series to be sliced further by some attributes (and perhaps with totals as well)
 * then create two-dim result resembling a pivot table:
 *
 * -  slices are in rows (first dim)
 * -  measures & scoping attributes will be in columns (second dim)
 */
function seriesAndSlicesDim(
    seriesBy: IAttributeOrMeasure[],
    slices: IAttribute[],
    totals: ITotal[],
): IDimension[] {
    const firstDimItems: DimensionItem[] = slices.map(attributeLocalId);
    firstDimItems.push(...totals);

    return newTwoDimensional(
        firstDimItems,
        seriesBy.filter(isAttribute).map(attributeLocalId).concat(MeasureGroupIdentifier),
    );
}

/**
 * Given execute props, this will prepare execution to send to backend.
 *
 * @param options - create execution options
 * @internal
 */
export function createExecution(options: CreateExecutionOptions): IPreparedExecution {
    const {
        backend,
        workspace,
        seriesBy,
        slicesBy = [],
        filters = [],
        sortBy = [],
        totals = [],
        componentName = "Execution",
    } = options;
    invariant(
        backend && workspace,
        "backend and workspace must be either specified explicitly or be provided by context",
    );

    const dimensions = isEmpty(slicesBy)
        ? seriesOnlyDim(seriesBy)
        : seriesAndSlicesDim(seriesBy, slicesBy, totals);

    return backend
        .withTelemetry(componentName, options)
        .workspace(workspace)
        .execution()
        .forItems(seriesBy.concat(slicesBy), filters)
        .withSorting(...sortBy)
        .withDimensions(...dimensions);
}
