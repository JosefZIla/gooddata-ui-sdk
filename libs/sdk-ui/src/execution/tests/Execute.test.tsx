// (C) 2019-2022 GoodData Corporation
import React from "react";
import { render } from "@testing-library/react";
import { dummyBackend, dummyBackendEmptyData } from "@gooddata/sdk-backend-mockingbird";
import { createDummyPromise } from "../../base/react/tests/toolkit";
import { DataViewFacade } from "../../base/results/facade";
import { IAnalyticalBackend } from "@gooddata/sdk-backend-spi";
import { Execute, IExecuteProps } from "../Execute";
import { ReferenceMd } from "@gooddata/reference-workspace";
import { newAttributeSort, newPositiveAttributeFilter, newTotal } from "@gooddata/sdk-model";
import { createExecution, CreateExecutionOptions } from "../createExecution";
import { LoadingComponent } from "../../base/react/LoadingComponent";
import { IExecuteErrorComponent } from "../interfaces";
import { setupComponent } from "../../base/tests/testHelper";

const DummyBackendEmptyData = dummyBackendEmptyData();
const makeChild = () => jest.fn((_) => <div />);
const renderDummyExecutor = (
    child: jest.Mock<JSX.Element>,
    props: Omit<IExecuteProps, "backend" | "workspace" | "children" | "seriesBy"> = {},
    backend: IAnalyticalBackend = DummyBackendEmptyData,
) => {
    return setupComponent(
        <Execute backend={backend} workspace={"testWorkspace"} seriesBy={[ReferenceMd.Amount]} {...props}>
            {child}
        </Execute>,
    );
};

describe("Execute", () => {
    it("should start loading immediately and inject isLoading prop", () => {
        const child = makeChild();
        renderDummyExecutor(child);

        expect(child).toHaveBeenCalledWith({
            isLoading: true,
            error: undefined,
            result: undefined,
            reload: expect.any(Function),
        });
    });

    it("should not start loading immediately if loadOnMount is set to false", () => {
        const child = makeChild();
        renderDummyExecutor(child, { loadOnMount: false });

        expect(child).toHaveBeenCalledWith({
            isLoading: false,
            error: undefined,
            result: undefined,
            reload: expect.any(Function),
        });
    });

    it("should stop loading when execution is resolved and inject data view facade", async () => {
        const child = makeChild();
        renderDummyExecutor(child);
        await createDummyPromise({ delay: 100 });

        expect(child).toHaveBeenCalledWith({
            isLoading: false,
            error: undefined,
            result: expect.any(DataViewFacade),
            reload: expect.any(Function),
        });
    });

    it("should stop loading when execution fails", async () => {
        const child = makeChild();
        renderDummyExecutor(child, {}, dummyBackend());
        await createDummyPromise({ delay: 100 });

        expect(child).toHaveBeenCalledWith({
            isLoading: false,
            error: expect.any(Error),
            result: undefined,
            reload: expect.any(Function),
        });
    });

    it("should start loading after invoking injected reload function", async () => {
        const child = jest.fn(({ reload }) => <button onClick={reload}>Reload</button>);
        const { getByText, user } = renderDummyExecutor(child, { loadOnMount: false });
        await user.click(getByText("Reload"));

        expect(child).toHaveBeenCalledWith({
            isLoading: false,
            error: undefined,
            result: undefined,
            reload: expect.any(Function),
        });
        expect(child).toHaveBeenCalledWith({
            isLoading: true,
            error: undefined,
            result: undefined,
            reload: expect.any(Function),
        });
    });

    it("should invoke onLoadingStart, onLoadingChanged and onLoadingFinish events", async () => {
        const child = makeChild();
        const onLoadingStart = jest.fn();
        const onLoadingChanged = jest.fn();
        const onLoadingFinish = jest.fn();

        renderDummyExecutor(child, {
            onLoadingChanged,
            onLoadingFinish,
            onLoadingStart,
        });

        await createDummyPromise({ delay: 100 });

        expect(onLoadingStart).toBeCalledTimes(1);
        expect(onLoadingChanged).toBeCalledTimes(2);
        expect(onLoadingFinish).toBeCalledTimes(1);
    });

    it("should invoke onError when execution fails with a NoDataError without a DataView", async () => {
        const child = makeChild();
        const onLoadingStart = jest.fn();
        const onLoadingChanged = jest.fn();
        const onLoadingFinish = jest.fn();
        const onError = jest.fn();

        renderDummyExecutor(
            child,
            {
                onError,
                onLoadingChanged,
                onLoadingFinish,
                onLoadingStart,
            },
            dummyBackend({ raiseNoDataExceptions: "without-data-view" }),
        );

        await createDummyPromise({ delay: 100 });

        expect(onError).toBeCalledTimes(1);
        expect(onLoadingStart).toBeCalledTimes(1);
        expect(onLoadingChanged).toBeCalledTimes(2);
        expect(onLoadingFinish).not.toBeCalled();
    });

    it("should NOT invoke onError when execution fails with a NoDataError with a DataView", async () => {
        const child = makeChild();
        const onLoadingStart = jest.fn();
        const onLoadingChanged = jest.fn();
        const onLoadingFinish = jest.fn();
        const onError = jest.fn();

        renderDummyExecutor(
            child,
            {
                onError,
                onLoadingChanged,
                onLoadingFinish,
                onLoadingStart,
            },
            dummyBackend({ raiseNoDataExceptions: "with-data-view" }),
        );

        await createDummyPromise({ delay: 100 });

        expect(onError).not.toBeCalled();
        expect(onLoadingStart).toBeCalledTimes(1);
        expect(onLoadingChanged).toBeCalledTimes(2);
        expect(onLoadingFinish).toBeCalledTimes(1);
    });

    it("should render LoadingComponent", async () => {
        const child = makeChild();
        const { queryByText } = renderDummyExecutor(child, {
            LoadingComponent: () => <div>MOCKED LOADING</div>,
        });

        expect(queryByText("MOCKED LOADING")).toBeInTheDocument();
    });

    it("should render ErrorComponent, when execution fails", async () => {
        const ErrorComponent: IExecuteErrorComponent = () => <div>MOCKED ERROR</div>;
        const child = makeChild();
        const { queryByText } = renderDummyExecutor(
            child,
            {
                ErrorComponent,
            },
            dummyBackend(),
        );

        await createDummyPromise({ delay: 100 });

        expect(queryByText("MOCKED ERROR")).toBeInTheDocument();
    });

    it("should not call children function without result, when both Loading & ErrorComponent are provided", async () => {
        const ErrorComponent: IExecuteErrorComponent = () => <div />;
        const child = makeChild();
        renderDummyExecutor(child, {
            LoadingComponent,
            ErrorComponent,
        });

        expect(child).not.toHaveBeenCalledWith(
            expect.objectContaining({
                result: undefined,
            }),
        );

        await createDummyPromise({ delay: 100 });

        expect(child).toHaveBeenCalledWith(
            expect.objectContaining({
                result: expect.any(DataViewFacade),
            }),
        );
    });

    it("should throw if neither seriesBy not slicesBy have any elements", async () => {
        const { queryByText } = render(
            <Execute backend={DummyBackendEmptyData} workspace={"testWorkspace"}>
                {({ error }) => {
                    if (error) {
                        return <div className="my-error">ERROR</div>;
                    }
                    return null;
                }}
            </Execute>,
        );

        await createDummyPromise({ delay: 100 });

        expect(queryByText("ERROR")).toBeInTheDocument();
    });
});

describe("createExecution", () => {
    const Scenarios: Array<[string, Partial<IExecuteProps>]> = [
        ["unscoped series only", { seriesBy: [ReferenceMd.Amount] }],
        ["scoped series only", { seriesBy: [ReferenceMd.Amount, ReferenceMd.Region] }],
        [
            "unscoped series with slicing",
            { seriesBy: [ReferenceMd.Amount], slicesBy: [ReferenceMd.Product.Name] },
        ],
        [
            "scoped series with slicing",
            { seriesBy: [ReferenceMd.Amount, ReferenceMd.Region], slicesBy: [ReferenceMd.Product.Name] },
        ],
        [
            "scoped series with slicing and filter",
            {
                seriesBy: [ReferenceMd.Amount, ReferenceMd.Region],
                slicesBy: [ReferenceMd.Product.Name],
                filters: [newPositiveAttributeFilter(ReferenceMd.Product.Name, ["CompuSci"])],
            },
        ],
        [
            "scoped series with slicing and sortBy",
            {
                seriesBy: [ReferenceMd.Amount, ReferenceMd.Region],
                slicesBy: [ReferenceMd.Product.Name],
                sortBy: [newAttributeSort(ReferenceMd.Product.Name, "desc")],
            },
        ],
        [
            "scoped series with slicing and totals",
            {
                seriesBy: [ReferenceMd.Amount, ReferenceMd.Region],
                slicesBy: [ReferenceMd.Product.Name],
                totals: [newTotal("sum", ReferenceMd.Amount, ReferenceMd.Product.Name)],
            },
        ],
        ["unscoped series only with only attributes (RAIL-4265)", { seriesBy: [ReferenceMd.Product.Name] }],
    ];

    it.each(Scenarios)("should create correct definition for %s", (_desc, props) => {
        const input: IExecuteProps = {
            backend: DummyBackendEmptyData,
            workspace: "testWorkspace",
            seriesBy: props.seriesBy!,
            children: makeChild(),
            ...props,
        };

        const execution = createExecution(input as CreateExecutionOptions);

        expect(execution.definition).toMatchSnapshot();
    });
});
