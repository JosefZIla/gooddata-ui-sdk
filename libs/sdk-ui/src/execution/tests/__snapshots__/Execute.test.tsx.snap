// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createExecution should create correct definition for scoped series only 1`] = `
Object {
  "attributes": Array [
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
  ],
  "buckets": Array [],
  "dimensions": Array [
    Object {
      "itemIdentifiers": Array [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": Array [],
  "measures": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": Object {},
  "sortBy": Array [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution should create correct definition for scoped series with slicing 1`] = `
Object {
  "attributes": Array [
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": Array [],
  "dimensions": Array [
    Object {
      "itemIdentifiers": Array [
        "a_label.product.id.name",
      ],
    },
    Object {
      "itemIdentifiers": Array [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": Array [],
  "measures": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": Object {},
  "sortBy": Array [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution should create correct definition for scoped series with slicing and filter 1`] = `
Object {
  "attributes": Array [
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": Array [],
  "dimensions": Array [
    Object {
      "itemIdentifiers": Array [
        "a_label.product.id.name",
      ],
    },
    Object {
      "itemIdentifiers": Array [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": Array [
    Object {
      "positiveAttributeFilter": Object {
        "displayForm": Object {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "in": Object {
          "values": Array [
            "CompuSci",
          ],
        },
      },
    },
  ],
  "measures": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": Object {},
  "sortBy": Array [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution should create correct definition for scoped series with slicing and sortBy 1`] = `
Object {
  "attributes": Array [
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": Array [],
  "dimensions": Array [
    Object {
      "itemIdentifiers": Array [
        "a_label.product.id.name",
      ],
    },
    Object {
      "itemIdentifiers": Array [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": Array [],
  "measures": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": Object {},
  "sortBy": Array [
    Object {
      "attributeSortItem": Object {
        "attributeIdentifier": "a_label.product.id.name",
        "direction": "desc",
      },
    },
  ],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution should create correct definition for scoped series with slicing and totals 1`] = `
Object {
  "attributes": Array [
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.owner.region",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.owner.region",
      },
    },
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": Array [],
  "dimensions": Array [
    Object {
      "itemIdentifiers": Array [
        "a_label.product.id.name",
      ],
      "totals": Array [
        Object {
          "attributeIdentifier": "a_label.product.id.name",
          "measureIdentifier": "m_aangOxLSeztu",
          "type": "sum",
        },
      ],
    },
    Object {
      "itemIdentifiers": Array [
        "a_label.owner.region",
        "measureGroup",
      ],
    },
  ],
  "filters": Array [],
  "measures": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": Object {},
  "sortBy": Array [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution should create correct definition for unscoped series only 1`] = `
Object {
  "attributes": Array [],
  "buckets": Array [],
  "dimensions": Array [
    Object {
      "itemIdentifiers": Array [
        "measureGroup",
      ],
    },
  ],
  "filters": Array [],
  "measures": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": Object {},
  "sortBy": Array [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution should create correct definition for unscoped series only with only attributes (RAIL-4265) 1`] = `
Object {
  "attributes": Array [
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": Array [],
  "dimensions": Array [
    Object {
      "itemIdentifiers": Array [
        "a_label.product.id.name",
      ],
    },
  ],
  "filters": Array [],
  "measures": Array [],
  "postProcessing": Object {},
  "sortBy": Array [],
  "workspace": "testWorkspace",
}
`;

exports[`createExecution should create correct definition for unscoped series with slicing 1`] = `
Object {
  "attributes": Array [
    Object {
      "attribute": Object {
        "displayForm": Object {
          "identifier": "label.product.id.name",
          "type": "displayForm",
        },
        "localIdentifier": "a_label.product.id.name",
      },
    },
  ],
  "buckets": Array [],
  "dimensions": Array [
    Object {
      "itemIdentifiers": Array [
        "a_label.product.id.name",
      ],
    },
    Object {
      "itemIdentifiers": Array [
        "measureGroup",
      ],
    },
  ],
  "filters": Array [],
  "measures": Array [
    Object {
      "measure": Object {
        "definition": Object {
          "measureDefinition": Object {
            "item": Object {
              "identifier": "aangOxLSeztu",
              "type": "measure",
            },
          },
        },
        "localIdentifier": "m_aangOxLSeztu",
      },
    },
  ],
  "postProcessing": Object {},
  "sortBy": Array [],
  "workspace": "testWorkspace",
}
`;
